### PR TITLE
Make ftp EC2 instance

### PR DESCRIPTION
**Pull request by rija**
# Pull request for issue: #1634, #1824, #1829, #1831

This is a pull request for the following functionalities:

* [x] Terraform to generate EC2 instance for file serving purpose (e.g: FTP, SFTP, TUS)
* [x] Update Ansible bootstrap to have the new FTP private and public IPs sent to Gitlab variables
* [x] New playbook to configure the new EC2 instance with the basics (fail2ban, docker, prometheus)
* [x] Pipeline change to push Pure-FTP server to the EC2 instance
* [x] Manually mount EFS on to the FTP server
* [x] Make possible to create user account that use EFS
* [x] Allow Set up of domain name for file server (EIP, Gitlab variable for hostname)
* [x] Deploy Pure-FTPD docker-compose
* [x] Update docs
* [x] Changed Database S3 backup time to be outside UK/HK office hours
* [x] Fix an issue where CSP was preventing frontend script to load

## How to test?

1. Setup your AWS staging environment as described in the developer documentation: 
`docs/developers/SETUP_AWS_ENVIRONMENTS.md`

2. Push this branch to your Gitlab pipeline

3. Follow this documentation to the very end to setup and test the FTP server:
`docs/sop/DEPLOY_AND_CONFIGURE_PUBLIC_FTP_SERVER.md`

>**Note**: Make sure you have read the "Any issues with implementation?" section below before starting testing

## How have functionalities been implemented?

Deploy Pure-FTPd on a new, distinct EC2 instance (so FTP activity doesn't steal bandwidth from the web server, and to protect the web server from the security liabilities of FTP protocol)

We use Pure-FTPd, because:
 * That's already what we use in the first iteration of FUW and we can reuse  the existing Pure-FTPd project
 * It's very secure
 * It's a small application
 * It's very simple to configure (especially programmatically)
 * It's very well documented
 * It's open-source and free

Having a new EC2 instance will be necessary for the File Upload Wizard re-implementation project anyways (to host the TUSd server and the FTP server), so the work in this PR both covers an immediate need (CNGB dropbox replacement) and being reusable as a foundation for the upcoming FUW functionalities, while keeping user credentials under our control (which would not be the case if using alternative based on AWS managed services - which have other issues anyways).

### Infrastructure provisioning and configuration

The Terraform module `ops/infrastructure/modules/aws-instance/aws-instance.tf` was modified to be more generic, so we can re-use it multiple times in `ops/infrastructure/terraform.tf` to have multiple EC2 instances on private subnets  ; for now: one for the web server, one for the file server.

There is an update to the `ops/infrastructure/bootstrap_playbook.yml` bootstrapping Ansible playbook so to be more generic, so we can call it multiple times, once per EC2 instance, in order to save the AWS IP addresses to Gitlab variables.

Various changes were made to existing playbooks and to the main Terraform configuration file`ops/infrastructure/terraform.tf` so the resource names and variables have a more consistent, parameter-friendly naming.

There is a new Ansible playbook to configure the new EC2 instance, `ops/infrastructure/files_playbook.yml`, that will first perform the same basic configuration tasks needed on all EC2 intances:
* Setup Fail2ban
* Install Docker
* Install Prometheus node

And in addition, it will:

* Install EFS-utils 
* Create the `/share/dropbox` and `/share/config` mount points directories
* Generate a Docker compose file for the FTP server from template `ops/configuration/file-server/docker-compose.yml.j2`

#### Project directory and docker file

We are able to make use of the existing project at `fuw/docker-pure-ftpd`. 
The Dockerfile there can be reused almost as-is, and the project directory can be home to the Gitlab configuration files for this project. The change to the Dockerfile is to create a new system user associated to the ftp account to be used for dropbox as the `downloader` and `uploader` system user were created for fuw. So there is now a `dropboxuser` system user for the use case at hand and this is reflected in the docs relating to creating user account: `docs/sop/DEPLOY_AND_CONFIGURE_PUBLIC_FTP_SERVER.md`.

#### Configuration and deployment

There is a  docker-compose file written on the EC2 instance as `/home/centos/docker-compose.yml` by the Ansible playbook that will allow operators to control the lifecycle of the Pure-FTPd on the server.
It also sets up Pure-FTPd commands to use the EFS mounted directories using Docker bind-mounting

>**Note**: the EFS still need to mounted manually for now, as documented in `docs/sop/DEPLOY_AND_CONFIGURE_PUBLIC_FTP_SERVER.md`

The Pure-FTPd software is deployed onto the new EC2 using Gitlab pipeline. To that effect there are new Gitlab configuration files added into `fuw/docker-pure-ftpd`. 
Their usage is explained in `docs/sop/DEPLOY_AND_CONFIGURE_PUBLIC_FTP_SERVER.md`

## Any issues with implementation?

* The `aws-instance` module is not as generic as it could be. 
  * In particular the list of open ports should be made more flexible, but it would require more complex changes than necessary for the current context (migration out of CNGB) 
  * and for the moment, the new file server EC2 instance will re-use the same instance type as defined for the web server
* Our docs are still a bit of a mess, but I don't have time to fully refactor it. We will need to gradually follow guidelines from [https://diataxis.fr](https://diataxis.fr)
* AWS only allow 5 EIPs per region and account, but with three EC2 instances and two environments a were are now past that limit. See below for short-term remediation.

#### On developers AWS environment

>**Note**: It normally should not be much of a problem since we do not use the live environment often on development (unless we are testing something related to release tag).

Make sure you only have only one environment running at any given time (usually staging).
If you want to switch to live, run `terraform destroy` in `/ops/infratructure/envs/staging`.
But before destroying staging, make sure the build pipeline for a **tag branch** has completed all the production build and deploy jobs on staging so you can manually trigger live build and deploy once the live environment is provisioned.

Then, In the AWS console screen listing the EIPs in your region, you can rename the EIPs straight from the list to match the environment you currently want to build (staging or live), by replacing `-staging-` with `-live-` or vice versa, so that you don't need more than 3 EIPs overall.
Finally you can build the live environment.

#### On Upstream

On Upstream the issue is quite a problem as we need all endpoints on staging and live to be accessible all the time.
So, I have requested to AWS an increase of the quota for "EC2-VPC Elastic IPs" in the "ap-east-1" region from 5 to 10.
See: https://ap-east-1.console.aws.amazon.com/servicequotas/home/services/ec2/quotas/L-0263D0A3

## Any changes to automated tests?

N/a

## Any changes to documentation?

* Updated production deploy SOP, `docs/sop/PRODUCTION_DEPLOY.md`, to take into account the new file server
* Added developer deploy doc for creating AWS environment: `docs/developers/SETUP_AWS_ENVIRONMENTS.md`
* Added to `docs/variables.md` new variables needed by the new service
* Updated CI/CD docs, `docs/SETUP_CI_CD_PIPELINE.md`, to take into account the new file server

## Any technical debt repayment?

Remove old unused docs, and make a directory for the docs relevant to developers activities

## Any improvements to CI/CD pipeline?

* Made the docs more up-to-date
* Enabled the flexibility to make it easier to provision more EC2 instances on private subnets (will need that too when we implement blue/green deployment in #1803)

### CODE REVIEW

File: docs/developers/SETUP_AWS_ENVIRONMENTS.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Setup the developers AWS environments
+
+## Prerequisite
+
+* Have a Gitlab account that has been added to GigaScience project
+* Have a project under "Gigascience/forks" named "\<github handle\>-gigadb-website
+* Have the domain names registered and DNS A record pointing to the following EIPs (which also must be created):
+
+| endpoint | environment | domain | EIP | 
+| -- | -- | -- |
```
Comment (pli888) [inline docs/developers/SETUP_AWS_ENVIRONMENTS.md]: Need to add another column on this line in order for this markdown table to be properly rendered.

Comment (rija) [reply docs/developers/SETUP_AWS_ENVIRONMENTS.md]: fixed

**Review by kencho51 — APPROVED** [review]
Hi @rija,

**prasie:** I was able to follow steps in `docs/developers/SETUP_AWS_ENVIRONMENTS.md` and `ocs/sop/DEPLOY_AND_CONFIGURE_PUBLIC_FTP_SERVER.md` to spin up files server, and ncftpput/ncftpget a file as user0.
**praise:** Good to have `SETUP_AWS_ENVIRONMENTS.md`, so we dont need to specify these standard steps in each PR.
**praise:** Change the rds backup time from utc 1100 to 2000, which would not cause blockage to most of the users if there is a downtime.
**question(non-blocking):** The aws efs utils is installed in the ftp server using the `files_playbook` which extend the `aws-instance` module. But in this draft PR https://github.com/gigascience/gigadb-website/pull/1828, the `user_data` would be introduced the modules to install the aws efs utils during the `terraform apply` step. So I would like to know which way, ansible or terraform, is better in installing the aws efs uitls?
**thought(non-blocking):** Glad that the solution for solving the limited aws eip quota problem is provided, the workaround is manageable, but also very laborious. Is it better to increase the eip quota to 6 for developer, 3 for staging and 3 for live?  The request can be made at [here](https://console.aws.amazon.com/servicequotas/), just to have an extra eip should not incur too much cost.

Overall LGTM, happy to approve.

**Review by pli888 — APPROVED** [review]
**praise**: I could instantiate an EFS and mount it onto a FTP server running on an EC2 instance.
**praise**: I could log into FTP server, and upload, download and change directories using `ncftpput`
**praise**: Code looks good to me.
**praise**: Database backup time has been changed to a less intrusive UK and HK time.

I approve this PR.

**Pull request merged**